proposals/0000-decompose-newtypes.rst

```diff
@@ -0,0 +1,192 @@
+  H is not a newtype or rs(i) is representational
+  ---------------------------------------------------------------------------------------    CO-NTH
+     G |- Nth(i) G : taus(i) ~rs(i) sigmas(i)
+
+Note the extra "or rs(i) is representational".  That says that if we know ``(T ty1) ~R (T ty2)``
+then we know ``ty1 ~R ty2`` *if ``T``'s argument has representational role*. When is that?
+Here's an example::
+
+  type family F
+  newtype N a b c = MkN ([a], F c)
+
+  -- Inferred roles    a            b      c
+  -- type role N representational phantom nominal
+
+Here
+* ``b`` has phantom role: it is not even mentioned in the RHS
+* ``c`` has nominal rule: it appears in the RHS but only in the argument of a type family
+* ``a`` has reprsentational role: it appears in the RHS, and *not* under a type family
+
+Representational role means "if you know the RHS type, then you know ``a``"
+
+Change 2: user role signatures
+::::::::::::::::::::::::::::::
+
+In GHC today, the user can declare explicit roles::
+
+  newtype P a = MkP Int
+  type role P representational
+
+Here the argument of ``P`` is not mentioned in its RHS, so its role would be inferres as ``phantom``.  But GHC today allows you to override a phantom role
+to ``representational``.
+
+If we continue to allow this, the new CO-NTH rule would be unsound.  Clearly, knowing ``(P t1) ~R (P t2)`` does **not** imply that ``t1 ~R t2``.
+
+Hence change 2: **In a newtype declaration, we do not allow a user to give a representational role for a phantom argument.**  Any such
+attempt would simply be rejected.
+
+With that change, CO-NTH is sound.
+
+It is perfectly OK to specify a ``nominal`` role for a ``phantom`` argument, however.  Thus, returning to the example ``N`` in the
+"Change 1" section:
+
+* ``type role N representational nominal nominal`` is OK: we have changed ``phantom`` into ``nominal``.
+* ``type role N nominal phantom nominal`` is OK: we have changed ``representational`` into ``nominal``.
+* ``type role N representational reprsentational nominal`` is BAD: we have changed ``phantom`` into ``representational``.
+
+So far as role ascription is concerned, ``representational`` is the most informative role of the three: it guarantees that the
+type variable appears, in at least guaranteed-injective position in the RHS.
+
+Effect and Interactions
+-----------------------
+
+With the new rule, we can decompose ``(T s1 s2 s3) ~R (T t1 t2 t3)``,
+where ``T`` is a newtype, if *all three parameters are at representational role*.
+That strengthens type inference without introducing incompleteness.
+
+Moreover, arguments with representational role are very common: they are arguments that
+are mentioned, and not under type families.  The ``IO`` example above is a case in point.
+
+
+Costs and Drawbacks
+-------------------
+
+Implementation is easy.
+
+The only cost is Change 2: you can't give a user-ascribed representational role to a phantom argument.
```

Comment:
I found

```haskell
newtype UnliftedArray# (a :: TYPE 'UnliftedRep) = UnliftedArray# ArrayArray#
type role UnliftedArray# representational
```

in `primitive-unlifted`. (`primitive` uses `data` so there is no newtypes).

AFAICT, if this change is accepted, the `primitive-unlifted` interface will be worse, phantom role would be wrong, and nominal is unnecessarily strong. Using `data` is not an option (as then it won't be `primitive-unlifted`).

---

Another, example is 

```haskell
-- | Type-indexed list
newtype TypedList (f :: (k -> Type)) (xs :: [k]) = TypedList [Any]
  deriving (Typeable)
type role TypedList representational representational
```

from `dimensions`. Similar type occurs many times on hackage (`vinyl`'s `Rec`, `large-records`' `Rec`), but these use `TypedList [f Any]` like encoding, so `f` is actually not-phantom. Similarly `some` package has `newtype Some f = Some (f Any)`.

However, note the `xs` is also marked as representational. And it makes sense. One should be able to `coerce` `TypedList Maybe '[Int]` to `TypedList Maybe `[Down Int]`. 

---

[`data-elevator`](https://hackage.haskell.org/package/data-elevator-0.1.0.1/docs/Data-Elevator.html) defines

```haskell
type role Strict representational
newtype Strict (a :: LiftedType) where
  Strict_ :: Any @UnliftedType -> Strict a
```

Reply:
`prim-unlifted` is kind of a mess. The `BoxedRep` stuff was supposed to fix it, but I don't know if it can do so yet.

Reply:
@treeowl it's still an example of `newtype WithSafeInterface a = Unsafe Any` pattern. These are newtypes for a reason. Maybe GHC should have `type family AnyOne (a :: k) :: l where`.